pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +223,238 @@ pub mod pallet {
 			todo!()
 		}
 
-		// TODO: not sure if we really need that. Check that out with Raphael.
 		#[pallet::call_index(6)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn start_refund(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Self::do_start_refund(pool_id, Some(&who))
 		}
 
 		#[pallet::call_index(7)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn force_start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn force_start_refund(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Self::do_start_refund(pool_id, None)
 		}
 
-		// todo: check if we really need that tx.
 		#[pallet::call_index(8)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn destroy_accounts(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn refund_account(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			account: AccountIdLookupOf<T>,
+			asset_idx: u32,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let who = T::Lookup::lookup(account)?;
+
+			let pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_refunding(), Error::<T>::NotRefunding);
+
+			// get asset id from linked assets vector
+			let asset_id: &FungiblesAssetIdOf<T> = pool_details
+				.bonded_currencies
+				.get(asset_idx.saturated_into::<usize>())
+				.ok_or(Error::<T>::IndexOutOfBounds)?;
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			// nothing to distribute; refunding is complete, user should call start_destroy
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::ZeroAmount);
+
+			// TODO: remove any existing locks on the account prior to burning
+
+			// With amount = max_value(), this trait implementation burns the reducible balance on the account and returns the actual amount burnt
+			let burnt = T::Fungibles::burn_from(
+				asset_id.clone(),
+				&who,
+				Bounded::max_value(),
+				Precision::BestEffort,
+				Fortitude::Force,
+			)?;
+
+			// no funds available to be burnt on account; nothing to do here
+			ensure!(!burnt.is_zero(), Error::<T>::ZeroAmount); // TODO: fail or return?
+
+			let sum_of_issuances = pool_details
```

Comment:
Iterating over all currencies will most likely always be part of mint, burn, and also refund operations. This is because all of them depend on the total supply in the pool's currencies, which we can only get from the balances pallet. And which will always require one lookup per currencies.

If we have to add the pool's number of linked currencies to the parameter set for benchmarking reasons, then that's probably the smallest price to pay.

What would a worst case calculation look like? Can that be based on the max_currencies constant for example?

Reply:
Yes, it's either taking the worst case as returned by the pallet constant, or provide the expected number of currencies stored and verify that condition as early as possible within the extrinsic. I mentioned this in a different PR and @Ad96el was thinking to go with the latter approach, which is a bit less user-friendly, especially if done via "raw" interactions, but should be fairly easily abstractable by dapps.

Reply:
I think that would also be my preference, depends a lot on the impact of worst-case benchmarking though. We should make sure to choose one approach throughout though.

Reply:
For one or two bonded currencies, this won't be an issue. The issue could arise, if I take the staking pallet as an example, once we get past thousdand, or more.

Reply:
But of course that also depends entirely on the HW used when benchmarking. If this pallet is to be used by other chains, this number is totally meaningless.

Reply:
I expect the max_currencies to be set to a two-digit number, but would stick to the latter approach chosen by @Ad96el for now. We can revisit when benchmarking.

Reply:
93f29503e5ccbeee4157ac08459d0a0eb92bb98e